Sources/NIO/FileHandle.swift

```diff
@@ -81,12 +81,43 @@ public final class NIOFileHandle: FileDescriptor {
 }
 
 extension NIOFileHandle {
+    /// `Mode` represents file access modes.
+    public struct Mode: OptionSet {
+        public let rawValue: Int32
+
+        public init(rawValue: Int32) {
+            self.rawValue = rawValue
+        }
+
+        /// Opens file for reading
+        public static let read = Mode(rawValue: O_RDONLY)
+        /// Opens file for writing
+        public static let write = Mode(rawValue: O_WRONLY)
+    }
```

Comment:
```suggestion
    /// `Mode` represents file access modes.
	    public struct Mode: OptionSet {
	        public let rawValue: UInt8
	
	        public init(rawValue: Int32) {
	            self.rawValue = rawValue
	        }
	
	        /// Opens file for reading
	        public static let read = Mode(rawValue: 1 << 0)
	        /// Opens file for writing
	        public static let write = Mode(rawValue: 1 << 1)
	    }
	    
	    internval var posixMode: CInt {
	        if self == [.read, .write] {
	            return O_RDWR
	       } else {
	           if self.contains(.read) {
	               return O_RDONLY
	           }
	           if self.contains(.write) {
	               return O_WRONLY
	           }
	           return 0
	      }
	  }
```

unfortunately, I don't think we can use the underlying POSIX values here. I don't think POSIX guarantees that `O_RDONLY | O_WRONLY == O_RDWR`. I think you will need to change this to something like the above which creates an OptionSet and then maps it to POSIX.

Reply:
Maybe then we don't need `OptionSet` at all? Especially given that users can add arbitrary values to it. What do we do in that case? Ignore user-provided values? Fail/crash if user specified something unexpected?

Reply:
@artemredkin but the `OptionSet` gives us the nice `[.read, .write]` syntax for free, right?

Reply:
Yes, it will, but how should we treat something like this: ```let mode: Mode = [.read, .init(rawValue: 127), .init(rawValue: UInt8(O_CREAT))]``` for example?

Reply:
well, they're illegal. We can ignore or crash. Sure, it's not really awesome but that's how `OptionSet`s work...

Reply:
if you have it too much you could also implement it from `ArrayLiteralConvertible`.

Reply:
Ignore or crash are runtime errors, wouldn't it be better to have compile-time safety here by using enums?

Reply:
@artemredkin Sure, it's just how `OptionSet`s work. I'm totally fine if you build a thing that doesn't use `OptionSet`s but I think it should probably be extensible. With an `enum` we can just about support `read` and `write`, by adding a third case `readAndWrite`. But adding anything else will be painful because 1) it breaks public API and 2) the number of cases will explode. For example if we need to support `O_APPEND` then it would be much cooler do support `[.read, .write, .append]` rather than having

```
enum Mode {
    case read
    case write
    case writeAndAppend
    case readAndWrite
    case readAndWriteAndAppend
}
```

just because there are so many things we need to support I'd like to have it extensible. The standard solution in Swift is `OptionSet` but as I said above, totally happy for you to not use `OptionSet` and make `[.read, .write]` work using `ArrayLiteralConvertible`.

Reply:
`open`'s manual states:
```
The argument flags must include one of the following access modes:
       O_RDONLY, O_WRONLY, or O_RDWR.
```
so those three are required, they are base modes, everything else is an option. So potential `O_APPEND` would be better set by `Flags`, in exact same manner as `O_CREAT`:
```swift
public struct Flags {
    public static let append = Flags(posixMode: O_APPEND, posixFlags: 0)
    public static let async = Flags(posixMode: O_ASYNC, posixFlags: 0)
    public static let tmp = Flags(posixMode: O_TMPFILE,  posixFlags: 0)
}

let file = openFile(path: "/some/file", mode: .writeOnly, flags: [.append, .async, .allowFileCreation()])
```
That would make it consistent with how we set `O_CREAT`. What do you think?

Reply:
@artemredkin well, but then there's `O_EVTONLY` which is in the same category as `O_RDONLY`, `O_WRONLY`, and `O_RDWR`

Reply:
Makes sense. Not sure `ArrayLiteralConvertible` looks nice, left as an `OptionSet` and added `precondition` to prevent passing in unsupported values